src/charm.py

```diff
@@ -8,35 +8,208 @@
+        """Handles certificate available events.
+
+        Finds the relation ID matching the CSR and forwards the received
+        certificate to that relation.
+        If a relation ID is not found, logs an error and ignores the event.
+
+        Args:
+            event (CertificateAvailableEvent): Event containing the certificate
+
+        Returns:
+            None
+        """
+        relation_id = self._get_relation_id_for_csr(event.certificate_signing_request)
+        if not relation_id:
+            logger.error(
+                "Could not find the relation for CSR: %s.",
+                event.certificate_signing_request,
+            )
+            return
+        self.certificates_requirers.set_relation_certificate(
+            certificate=event.certificate,
+            certificate_signing_request=event.certificate_signing_request,
+            ca=event.ca,
+            chain=event.chain,
+            relation_id=relation_id,
+        )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handles certificate invalidated events.
+
+        If the certificate is invalidated because it expired, ignore the event
+        and let the requirer handle it properly. Otherwise, calls the TLS
+        library to revoke the certificate to the requirer.
+
+        Args:
+            event (CertificateInvalidatedEvent): Event for invalidated certificate
+
+        Returns:
+            None
+        """
+        if event.reason == "expired":
+            return
+        self.certificates_requirers.remove_certificate(event.certificate)
+
+    def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEvent) -> None:
+        """Handles all certificates invalidated events.
+
+        Revokes all certificates.
+
+        Args:
+            event (AllCertificatesInvalidatedEvent): Event for all certificates invalidated
+
+        Returns:
+            None
+        """
+        self.certificates_requirers.revoke_all_certificates()
+
+    def _get_relation_id_for_csr(self, csr: str) -> Optional[int]:
+        """Finds the relation ID that sent the provided CSR.
+
+        This should return a single relation ID, otherwise it means multiple
+        applications requested the same certificate using the same private
+        key. In that case, we log an error and return None.
+
+        Args:
+            csr (str): Certificate Signing Request to search
+
+        Returns:
+            Relation ID (int) or None
+        """
+        requirers_csrs = self.certificates_requirers.get_requirer_csrs()
+        relation_ids: set[int] = set()
+        for requirer_csrs in requirers_csrs:
+            match requirer_csrs:
+                case {
+                    "relation_id": relation_id,
+                    "unit_csrs": [{"certificate_signing_request": str(unit_csr)}],
+                } if unit_csr == csr:  # type: ignore[has-type]
```

Comment:
I don't know about having a match statement with a single case :D
```suggestion
                unit_csrs = [entry["certificate_signing_request"] for entry in requirer_csr["unit_csrs"]]
                if requirer_csr["relation_id"] == relation_id and csr in unit_csrs:
```

Reply:
There is nothing wrong with a match statement with a single case, this is not a switch statement.

But I agree that the suggestion is simpler. I tweaked it a bit to use a generator instead of a list for performance reasons, and it does not impact readability.

Reply:
I've always imagined match as a pythonized switch statement but now that I think about it you're right :D